instrumentation/github.com/99designs/gqlgen/otelgqlgen/gqlgen.go

```diff
@@ -0,0 +1,162 @@
+
+package otelgqlgen
+
+import (
+	"context"
+	"os"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/99designs/gqlgen/graphql/handler/extension"
+
+	otelcontrib "go.opentelemetry.io/contrib"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+const (
+	tracerName         = "go.opentelemetry.io/contrib/instrumentation/github.com/99designs/gqlgen/otelgqlgen"
+	extensionName      = "OpenTelemetry"
+	complexityLimit    = "ComplexityLimit"
+	defaultServiceName = "GraphQLService"
+)
+
+type Tracer struct {
+	serviceName             string
+	complexityExtensionName string
+	tracer                  oteltrace.Tracer
+}
+
+var _ interface {
+	graphql.HandlerExtension
+	graphql.ResponseInterceptor
+	graphql.FieldInterceptor
+} = Tracer{}
+
+func (a Tracer) ExtensionName() string {
+	return extensionName
+}
+
+func (a Tracer) Validate(_ graphql.ExecutableSchema) error {
+	return nil
+}
+
+func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	ctx, span := a.tracer.Start(ctx, operationName(ctx), oteltrace.WithSpanKind(oteltrace.SpanKindServer))
+	defer span.End()
+	if !span.IsRecording() {
+		return next(ctx)
+	}
+
+	oc := graphql.GetOperationContext(ctx)
+
+	serviceName := os.Getenv("OTEL_SERVICE_NAME")
+	if serviceName == "" {
+		serviceName = defaultServiceName
+	}
+	span.SetAttributes(
+		semconv.ServiceNameKey.String(serviceName),
+		RequestQuery(oc.RawQuery),
+	)
+	complexityExtension := a.complexityExtensionName
+	if complexityExtension == "" {
+		complexityExtension = complexityLimit
+	}
+	complexityStats, ok := oc.Stats.GetExtension(complexityExtension).(*extension.ComplexityStats)
+	if !ok {
+		// complexity extension is not used
+		complexityStats = &extension.ComplexityStats{}
+	}
+
+	if complexityStats.ComplexityLimit > 0 {
+		span.SetAttributes(
+			RequestComplexityLimit(int64(complexityStats.ComplexityLimit)),
+			RequestOperationComplexity(int64(complexityStats.Complexity)),
+		)
+	}
+
+	span.SetAttributes(RequestVariables(oc.Variables)...)
+
+	return next(ctx)
+}
+
+func (a Tracer) InterceptField(ctx context.Context, next graphql.Resolver) (interface{}, error) {
+	fc := graphql.GetFieldContext(ctx)
+	ctx, span := a.tracer.Start(ctx,
+		fc.Field.ObjectDefinition.Name+"/"+fc.Field.Name,
+		oteltrace.WithSpanKind(oteltrace.SpanKindServer),
+	)
+	defer span.End()
+	if !span.IsRecording() {
+		return next(ctx)
+	}
+
+	span.SetAttributes(
+		ResolverPath(fc.Path().String()),
+		ResolverObject(fc.Field.ObjectDefinition.Name),
+		ResolverField(fc.Field.Name),
+		ResolverAlias(fc.Field.Alias),
+	)
+	span.SetAttributes(ResolverArgs(fc.Field.Arguments)...)
+
+	resp, err := next(ctx)
+
+	errList := graphql.GetFieldErrors(ctx, fc)
+	if len(errList) != 0 {
+		span.SetStatus(codes.Error, errList.Error())
+		span.SetAttributes(ResolverErrors(errList)...)
+	}
```

Comment:
shouldn't we set the span's status as error and record an error?

Reply:
i'm not sure that i catch the sence.

```
	errList := graphql.GetFieldErrors(ctx, fc)
	if len(errList) != 0 {
		span.SetStatus(codes.Error, errList.Error())
		span.RecordError(fmt.Errorf(errList.Error()))
		span.SetAttributes(ResolverErrors(errList)...)
	}
```

this will be correct in your opinion?

Reply:
Looks OK. And a unit test if possible 😉